### PR TITLE
Make a pass fixing Dataset API issues

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This fixes the following issues:
- should force compute="actors" when passing a stateful callable to map
- remove _spread_resource_prefix
- remove _move
- remove .pipeline
- promote to PublicAPI: fully_executed, stats
- DatasetPipeline: underscores for constants
- DatasetPipeline: add fetch_if_missing to schema
- DatasetPipeline: move stats to public api
- DatasetPipeline: remove foreach_dataset
- read_api: remove _spread_resource_prefix
- read_api: make tensor_column_schema supported
- compute: make ComputeStrategy DeveloperAPI

### CODE REVIEW

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -507,8 +507,6 @@ def random_shuffle(
         *,
         seed: Optional[int] = None,
         num_blocks: Optional[int] = None,
-        _spread_resource_prefix: Optional[str] = None,
```
Comment (jjyao) [inline python/ray/data/dataset.py:510]: @scv119 Have we confirmed with Uber that they no longer need `_spread_resource_prefix`?

File: python/ray/data/tests/test_dataset.py [hunk 2]
```diff
@@ -2964,83 +2968,6 @@ def range(n, parallelism=200):
...
-    ray.wait(blocks, num_returns=len(blocks), fetch_local=False)
-    location_data = ray.experimental.get_object_locations(blocks)
-    locations = []
-    for block in blocks:
-        locations.extend(location_data[block]["node_ids"])
-    assert set(locations) == {node1_id, node2_id}
-
-
-@pytest.mark.parametrize("use_spread_resource_prefix", [False, True])
-def test_parquet_read_spread(ray_start_cluster, tmp_path, use_spread_resource_prefix):
```
Comment (jjyao) [inline python/ray/data/tests/test_dataset.py:3000]: We should still keep these two tests for "use_spread_resource_prefix" false case (i.e. using "SPREAD" scheduling strategy).

File: python/ray/data/tests/test_dataset.py [hunk 3]
```diff
@@ -3055,9 +3052,6 @@ def test_parquet_read_spread_no_cpus(ray_start_cluster, tmp_path):
     def get_node_id():
```
Comment (jjyao) [inline python/ray/data/tests/test_dataset.py:3052]: this test can be deleted, seems it's not testing anything?

**Review by jjyao — COMMENTED** [review]
LGTM. I'll let @clarkzinzow to take another look.

File: python/ray/data/dataset.py (hunk 1 above)
Comment (clarkzinzow) [reply python/ray/data/dataset.py:510]: AFAIK we have not.

File: python/ray/data/dataset.py [hunk 4]
```diff
@@ -2756,11 +2733,24 @@ def fully_executed(self) -> "Dataset[T]":
...
+        until the underlying blocks are computed.
+
+        Time complexity: O(1)
+
+        Returns:
+            A list of references to this dataset's blocks.
+        """
+        return self._plan.execute().get_blocks()
+
     def _experimental_lazy(self) -> "Dataset[T]":
```
Comment (clarkzinzow) [inline python/ray/data/dataset.py:2754]: Should we de-experimentalize lazy evaluation before GA?

**Review by clarkzinzow — APPROVED** [review]
Nice! So happy to see `_spread_resource_prefix` go away. 🎉

File: python/ray/data/dataset.py (hunk 4 above)
Comment (ericl) [reply python/ray/data/dataset.py:2754]: Good question. I'm leaning towards no, given that we have read task fusion enabled at least, and most of the critical use cases involve pipelines. But we can revisit based on user needs.

**Pull request merged**